packages/flutter_tools/lib/src/android/gradle_errors.dart

```diff
@@ -669,3 +670,47 @@ final GradleHandledError sslExceptionHandler = GradleHandledError(
   },
   eventLabel: 'ssl-exception-tag-mismatch',
 );
+
+/// If an incompatible Java and Gradle versions error is caught, we expect an
+/// error specifying that the Java major class file version, one of
+/// https://javaalmanac.io/bytecode/versions/, is unsupported by Gradle.
+final RegExp _unsupportedClassFileMajorVersionPattern = RegExp(r'Unsupported class file major version\s+([0-9]+)');
+
+@visibleForTesting
+final GradleHandledError incompatibleJavaAndGradleVersionsHandler = GradleHandledError(
+  test: (String line) {
+    return _unsupportedClassFileMajorVersionPattern.hasMatch(line);
+  },
+  handler: ({
+    required String line,
+    required FlutterProject project,
+    required bool usesAndroidX,
+    required bool multidexEnabled,
+  }) async {
+    final File gradlePropertiesFile = project.directory
+        .childDirectory('android')
+        .childDirectory('gradle')
+        .childDirectory('wrapper')
+        .childFile('gradle-wrapper.properties');
+    final File gradleFile = project.directory
+        .childDirectory('android')
+        .childFile('build.gradle');
+    globals.printBox(
+      "${globals.logger.terminal.warningMark} Your project's Gradle and AGP version "
+      'is incompatible with the Java version that Flutter is using.\n\n'
+      'To fix this issue, first, check the Java version used by Flutter by '
+      'running `flutter doctor --verbose`.\n\n'
+      'Then, update the Gradle version specified in ${gradlePropertiesFile.path} '
+      'to be compatible with that Java version. '
+      'See the link below for more information on compatible Java/Gradle versions:\n'
+      'https://docs.gradle.org/current/userguide/compatibility.html#java\n\n'
+      'Then, ensure the AGP version specified in ${gradleFile.path} is '
```

Comment:
As I understand the compatibility table, upgrading your version of gradle should never make it incompatible with an unchanged version of AGP, is that correct? If so, do we need to add the step of upgrading AGP version to the suggestions here? 

The only case where they would have an incompatible version of gradle for their AGP version would be if it were already incompatible before changing the gradle version, and also if that is the case the existing error they will hit is pretty clear on both what the problem is and what the solution is.

Reply:
For context, this is what the error looks like if your gradle version is too low for your AGP version.
```
FAILURE: Build failed with an exception.

* Where:
Build file '/Users/mackall/development/BugTesting/114756/mysample/android/app/build.gradle' line: 24

* What went wrong:
A problem occurred evaluating project ':app'.
> Failed to apply plugin 'com.android.internal.version-check'.
   > Minimum supported Gradle version is 7.2. Current version is 6.7. If using the gradle wrapper, try editing the distributionUrl in /Users/mackall/development/BugTesting/114756/mysample/android/gradle/wrapper/gradle-wrapper.properties to gradle-7.2-all.zip
```

Reply:
There's another error that would catch this before this, but I don't think it can hurt to make sure that whatever they change it to is compatible? We assume they would not change it to something incompatible but it's possible, no?

Reply:
I think the error you get if gradle is too low for AGP is a useful error so we should probably leave that alone.

Reply:
Yes I do not think we need to add to the error message the android tooling gives